### ISSUE TITLE
testing/fstest: TestFS: relax check ReadDir based on perms

### ISSUE THREAD

**Issue by dolmen** [golang/go#63707]
## TL;DR

Drop one false positive case in [`testing/fstest.TestFS`](https://pkg.go.dev/testing/fstest#TestFS): it is ok for [`ReadDir`](https://pkg.go.dev/io/fs#FS.ReadDirFS) to return [`ErrPermission`](https://pkg.go.dev/io/fs#ErrPermission) if the [`fs.FileMode`](https://pkg.go.dev/io/fs#FileMode) of the [file](https://pkg.go.dev/io/fs#File) says some users are not allowed to read the directory.

## Context

With [`io/fs.FS`](https://pkg.go.dev/io/fs#FS), there is no way to select which permission bits ([`fs.FileMode`](https://pkg.go.dev/io/fs#FileMode)) apply when [reading a file](https://pkg.go.dev/io/fs#ReadFileFS). So fstest.TestFS has no way to enforce that given a FileMode's file/directory based on user/group/others permission bits, `Open`, [`ReadFile`](https://pkg.go.dev/io/fs#ReadFileFS) or [`ReadDir`](https://pkg.go.dev/io/fs#ReadDirFS) should trigger or not an [`fs.ErrPermission`](https://pkg.go.dev/io/fs#ErrPermission).

Currently any error (including [`ErrPermission`](https://pkg.go.dev/io/fs#ErrPermission)) on [`ReadDir`](https://pkg.go.dev/io/fs#FS.ReadDirFS) is always considered an error. However returning `fs.ErrPermission` is correct behaviour if the filesystem enforces `fs.FileMode` and the directory has read permissions bits off.

## Proposal

I propose that `TestFS` to ignore `ErrPermission` on [`ReadDir`](https://pkg.go.dev/io/fs#FS.ReadDirFS) if at least one *read* permission bit is not set: ignore `ErrPermission` (and silently skip other `ReadDir` checks) if `mode & 0444 != 0444`.

## Impact

This change will only drop some returned errors and will never raise more errors, so backward compatibility is preserved.

**Comment by dolmen** [comment]
Cc: @bcmills

**Comment by neild** [comment]
Permission bits aren't universally supported; if `TestFS` should ignore a permission error opening a file with no Unix read bits, shouldn't it ignore a permission error opening a Windows file with an ACL forbidding access?

**Comment by dolmen** [comment]
@josharian May I ask your opinion on this (as you were the author of #46776 as well as of various io/fs.FS implementations)?

**Comment by dolmen** [comment]
@neild The documentation for [`fs.FileMode`](https://pkg.go.dev/io/fs#FileMode) mentions:
> The nine least-significant bits are the standard Unix rwxrwxrwx permissions. The values of these bits should be considered part of the public API and may be used in wire protocols or disk representations

**Comment by seankhliao** [comment]
I don't think we should do anything, since this is a test helper, you should be in full control over your inputs and not call it on filesystems you don't have permissions for.
